src/__tests__/index.js

```diff
@@ -77,34 +79,69 @@ test("customWithCtrl", () => {
     value1: string,
     value2: string
   };
-  const Comp = withCtrl(
-    class extends Component<Props, State> {
-      state = {
-        value1: "state1",
-        value2: "state2"
-      };
-      render() {
-        return <div />;
-      }
-    },
-    {
-      mapPropsToState(props, state) {
-        const overrides = {};
-        if ("value" in props) {
-          const [value1, value2] = props.value.split(" ");
-          overrides.value1 = value1;
-          overrides.value2 = value2;
-        }
-        return overrides;
-      }
+
+  function mapPropsToState(props, state) {
+    const overrides = {};
+    if ("value" in props) {
+      const [value1, value2] = props.value.split(" ");
+      overrides.value1 = value1;
+      overrides.value2 = value2;
+    }
+    return overrides;
+  }
+
+  class Comp extends Component<Props, State> {
+    state = {
+      value1: "state1",
+      value2: "state2"
+    };
+    render() {
+      return (
+        <Ctrl component={this} mapPropsToState={mapPropsToState}>
```

Comment:
Part of the reason I liked this API:
```
<Ctrl controlledProps={ctrlProps}>
  {({ value }, setControlledState) => (
  // ...
  )}
</Ctrl>
```
Is that you can offload the state ownership to `Ctrl`, meaning you can write more stateless-functional-components. It also solves the problem of which props you want to be controlled.

That said, I like how you can call `this.setState` normally in this approach. It also feels pretty intuitive to imagine this component as a map function like `(props, state) => controlledState`.

Reply:
> It also solves the problem of which props you want to be controlled.

See my reply above to @Blasz. You can just do `this.state` on the host component.

> That said, I like how you can call this.setState normally in this approach. It also feels pretty intuitive to imagine this component as a map function like (props, state) => controlledState.

This is essentially the MO of the library. State ownership is the job of Unstated. That said, being able to pass in separate state and props means that you could compose this with unstated:

```js
render() {
  return (
    <Subscribe to={[MyContainer]}>
      {container => (
        <Ctrl props={this.props} state={container.state}>
          {mapped => (
            <StatelessFuncComponent {...mapped} />
          )}
        </Ctrl>
      )}
    </Subscribe>
  );
}
```

The thing I like about this approach is we've separated things into 3, composable, but separately tested parts:

1. Container / state
2. Default props / props / state mapping
3. The component

Part 3 is essentially the rest of render, but means that you can test it separately from your state. This is similar to the whole stateful / stateless thing we were doing before, in a way, but prevents you from having to duplicate anything.